### PR TITLE
new server that works

### PR DESCRIPTION
**Pull request by AviSananes**

### CODE REVIEW

File: GradeServer.c [hunk 1]
```diff
@@ -1,146 +1,216 @@
...
+#include <unistd.h>
+#include <pthread.h>
+#include <stdlib.h>
 
-int sockfd;
+#define NUM_THREADS 5
+#define MAX_TASKS 1000
 
 
-int loadUsersFromFile(User * users, const char* filename, int user_type) {
```
Comment (mikisegall) [inline GradeServer.c:7]: Why did you delete this function?

File: GradeServer.c [hunk 2]
```diff
@@ -1,146 +1,216 @@
...
+        // TODO - should we validate no additional arguments?
 
-void sigint_handler(int sig) {
-  printf("Received SIGINT, shutting down...\n");
+        case "Login":
+            login(arg1, arg2, response);
+            break;
 
-  // Close the socket and exit
-  close(sockfd);
```
Comment (mikisegall) [inline GradeServer.c:142]: In the old version we'd close the socket but you removed it - we don't need it?

File: GradeServer.c [hunk 3]
```diff
@@ -1,146 +1,216 @@
-#include <GradeServer.h>
-#include <ServerThread.h>
+#include <stdio.h>
```
Comment (mikisegall) [inline GradeServer.c:1]: All the includes were in the header file, why add it here?

File: GradeServer.c [hunk 4]
```diff
@@ -1,146 +1,216 @@
...
+{
+    pthread_mutex_lock(&task_mutex);
+
+    for (int i = 0; i < num_tasks; i++)
+    {
+        if (strcmp(tasks[i].message, task.message) == 0)
+        {
+            // Shift all tasks after the specified index one position to the left
+            for (int j = i; j < num_tasks - 1; j++)
+                tasks[j] = tasks[j + 1];
```
Comment (mikisegall) [inline GradeServer.c:193]: If we move to the "left", shouldn't it be `tasks[j] = tasks[j - 1];` instead of +?

File: GradeServer.c [hunk 5]
```diff
@@ -1,146 +1,216 @@
...
-    pthread_mutex_unlock(&task_list_mutex);
+task_t get_task(void)
+{
+    pthread_mutex_lock(&task_mutex);
+
+    while (num_tasks == 0)
+        pthread_cond_wait(&task_cond, &task_mutex);
+
+    // Get the next task from the list
+    task_t task = tasks[--num_tasks];
```
Comment (mikisegall) [inline GradeServer.c:177]: We reduce the num of tasks both in get + delete tasks. Wouldn't it create a bug?

File: GradeServer.c (hunk 1 above)
Comment (AviSananes) [reply GradeServer.c:7]: That's what I told you that I deleted and needed to add back, but my brain was on fire alreay

File: GradeServer.c (hunk 2 above)
Comment (AviSananes) [reply GradeServer.c:142]: Yeah I guess that we need to close this socket...

File: GradeServer.c (hunk 3 above)
Comment (AviSananes) [reply GradeServer.c:1]: Because I changed all the functions, we basically need a new header file

File: GradeServer.c (hunk 4 above)
Comment (AviSananes) [reply GradeServer.c:193]: First, it's the chat's function 😅
Second, I think by "shifting to the left" it means to "move the pointer to the right", so it's j=j+1

File: GradeServer.c (hunk 5 above)
Comment (AviSananes) [reply GradeServer.c:177]: You're right, we need to remove "delete_task". I missed it.
